### PR TITLE
Do not throw ObjectDisposeException from Socket.EndXyz methods

### LINKED ISSUE #61411: Socket.BeginConnect in .NET 6 throws an unobserved SocketException
### Description

After updating to .NET 6 we started seeing (separate users and software) unobserved `SocketException`.

After analyzing the logs, I believe it is being triggered here:

https://github.com/SteamRE/SteamKit/blob/2.4.0-Alpha.3/SteamKit2/SteamKit2/Networking/Steam3/TcpConnection.cs#L158

However it is hard to tell because the stacktrace doesn't really have useful information which could pin point the exact code.

The code that uses AwaitableSocketAsyncEventArgs is in `Socket.ConnectAsync`

@yaakov-h pointed out this:
> ValueTask<TResult>.AsTask is the only code path to there, and I don’t think SK2 knows about ValueTask, so that’s gotta be a runtime bug.

https://source.dot.net/#System.Private.CoreLib/ValueTask.cs,638

### Reproduction Steps

-

### Expected behavior

Exception should be catchable, or not happen in this manner?

### Actual behavior

```
Unhandled exception. System.AggregateException: A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread. (Operation canceled) (Operation canceled)
 ---> System.Net.Sockets.SocketException (125): Operation canceled
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.ThrowException(SocketError error, CancellationToken cancellationToken)
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.System.Threading.Tasks.Sources.IValueTaskSource.GetResult(Int16 token)
   at System.Threading.Tasks.ValueTask.ValueTaskSourceAsTask.<>c.<.cctor>b__4_0(Object state)
   --- End of inner exception stack trace ---
   at System.AggregateException.Handle(Func`2 predicate)
   at SteamDatabaseBackend.Bootstrapper.OnUnobservedTaskException(Object sender, UnobservedTaskExceptionEventArgs args) in /home/steamdb/deploy/Bootstrapper.cs:line 149
   at System.Threading.Tasks.TaskScheduler.PublishUnobservedTaskException(Object sender, UnobservedTaskExceptionEventArgs ueea)
   at System.Threading.Tasks.TaskExceptionHolder.Finalize()
```

### Regression?

I did not get this on .NET 5.

### Known Workarounds

_No response_

### Configuration

Debian 11 x64.

### Other information

_No response_

### PR DESCRIPTION
**Pull request by antonfirsov**
Fixes #61411 and cleans up some null-reference exception handling.

Considering (1) the amount of test changes (2) the fact that we are documenting ODE (3) complaints like #41585, I think it's reasonable to mark this change as breaking.

### CODE REVIEW

**Comment by karelz** [comment]
Triage: Throwing different exception in situation like this is NOT considered a breaking change. It is a technical breaking change, we do not plan to document, unless we see it has real-world impact on customers and causes confusion. Removing the labels.

File: src/libraries/System.Net.Sockets/tests/FunctionalTests/Accept.cs [hunk 1]
```diff
@@ -390,6 +382,52 @@ public sealed class AcceptSyncForceNonBlocking : Accept<SocketHelperSyncForceNon
...
+        public void AbortedByDispose_LeaksNoUnobservedExceptions()
+        {
+            RemoteExecutor.Invoke(static async () =>
+            {
+                var socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
+                socket.BindToAnonymousPort(IPAddress.Loopback);
+                socket.Listen(10);
+
+                bool unobservedThrown = false;
+                TaskScheduler.UnobservedTaskException += OnUnobservedException;
```
Comment (MihaZupan) [inline src/libraries/System.Net.Sockets/tests/FunctionalTests/Accept.cs:396]:
```suggestion
                TaskScheduler.UnobservedTaskException += (_, __) => unobservedThrown = true;
```

File: src/libraries/System.Net.Sockets/tests/FunctionalTests/Accept.cs [hunk 2]
```diff
@@ -390,6 +382,52 @@ public sealed class AcceptSyncForceNonBlocking : Accept<SocketHelperSyncForceNon
...
+                });
+
+                // Give some time for the Accept operation to start
+                await Task.Delay(30);
+
+                // Close the socket aborting Accept
+                socket.Dispose();
+
+                // Give chance for the unobserved exception to propagate. For some reason both waiting and enforcing Finalization is needed for this.
+                await Task.Delay(30);
```
Comment (MihaZupan) [inline src/libraries/System.Net.Sockets/tests/FunctionalTests/Accept.cs:419]:
```suggestion
                // Wait for the internal AcceptAsync Task to complete with the exception.
                await Task.Delay(30);

                // Ensure that the internal TaskExceptionHolder is finalized and the exception published to UnobservedTaskException.
```

File: src/libraries/System.Net.Sockets/tests/FunctionalTests/Accept.cs [hunk 3]
```diff
@@ -390,6 +382,52 @@ public sealed class AcceptSyncForceNonBlocking : Accept<SocketHelperSyncForceNon
...
+                bool unobservedThrown = false;
+                TaskScheduler.UnobservedTaskException += OnUnobservedException;
+
+                await Task.Run(() =>
+                {
+                    socket.BeginAccept(asyncResult =>
+                    {
+                        try
+                        {
+                            var accepted = socket.EndAccept(asyncResult);
```
Comment (MihaZupan) [inline src/libraries/System.Net.Sockets/tests/FunctionalTests/Accept.cs:404]:
```suggestion
                            socket.EndAccept(asyncResult);
```

**Review by MihaZupan — APPROVED** [review]

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs [hunk 4]
```diff
@@ -2428,22 +2404,18 @@ public IAsyncResult BeginReceive(IList<ArraySegment<byte>> buffers, SocketFlags
...
+        private static int EndSendReceive(IAsyncResult asyncResult, out SocketError errorCode)
         {
-            ThrowIfDisposed();
-
             if (TaskToApm.GetTask(asyncResult) is not Task<int> ti)
             {
-                throw new ArgumentException(null, nameof(asyncResult));
+                throw asyncResult is null ?
+                    new ArgumentNullException(nameof(asyncResult)) :
+                    new ArgumentException(null, nameof(asyncResult));
```
Comment (stephentoub) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:2418]: Is this a separate fix to throw a more specific exception?

Comment (stephentoub) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:2418]:
```suggestion
                ArgumentNullException.ThrowIfNull(asyncResult);
                throw new ArgumentException(null, nameof(asyncResult));
```

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs (hunk 4 above)
Comment (antonfirsov) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:2418]: Yes, this is to match the behavior of methods which simply delegate to `TaskToApm`:

https://github.com/dotnet/runtime/blob/73cf30dd4a00a1ee7b12a82a2cff99b1870a3c4e/src/libraries/Common/src/System/Threading/Tasks/TaskToApm.cs#L66-L71

**Comment by antonfirsov** [comment]
System.Runtime wasm failure is unrelated, outrloop pipeline seems to be broken.

**Pull request merged**